### PR TITLE
Move release-note labeling to Prow.

### LINKED ISSUE #4341: Changing release note block in PR body after release note is set does not update release note labels.
Currently the release-note-label munger ignores PRs that already have a release note added (i.e. have one of `release-note`, `release-note-none`, or `release-note-action-required`). This prevents the munger from changing the label if the PR body's release note block is updated.

Discussed here: #4295

### LINKED ISSUE #4505: Release note munger -> prow plugin
Right now we can say `/release-note-none`, but prow doesn't look at the PR body to determine what label should be applied. We should augment the `releasenote` plugin to handle pull request events and update the label. Then we can delete the release note munger.

### PR DESCRIPTION
**Pull request by cjwagner**
This PR moves the functionality of the `release-note-label` munger to Prow and improves it to fix #4341 .

It also removes the `/release-note` and `/release-note-action-required` commands from Prow because they are no longer needed and could conflict with the release note label specified by the `release-note` block in the PR body text. The `/release-note-none` command may still be used if the `release-note` block is empty.
fixes #4505 
/cc @stevekuznetsov @rmmh

### CODE REVIEW

**Comment by BenTheElder** [comment]
ref https://github.com/kubernetes/test-infra/issues/3331

File: prow/plugins/releasenote/releasenote.go [hunk 1]
```diff
@@ -29,98 +31,272 @@ import (
...
 	}
+	return gc.DeleteStaleComments(
+		org,
+		repo,
+		pr.Number,
+		func(c github.IssueComment) bool {
+			if c.User.Login != botName {
+				return false
+			}
+			if !strings.Contains(c.Body, releaseNoteBody) &&
```
Comment (rmmh) [inline prow/plugins/releasenote/releasenote.go]: shorten this to:

    return strings.Contains(c.Body, releaseNoteBody) || strings.Contains(c.Body, parentReleaseNoteBody) || c.Body == deprecatedReleaseNoteBody

**Comment by spxtr** [comment]
#4505

File: prow/plugins/releasenote/releasenote.go [hunk 2]
```diff
@@ -29,98 +31,272 @@ import (
...
 	}
-
-	releaseNoteRe               = regexp.MustCompile(`(?mi)^/release-note\s*$`)
-	releaseNoteNoneRe           = regexp.MustCompile(`(?mi)^/release-note-none\s*$`)
-	releaseNoteActionRequiredRe = regexp.MustCompile(`(?mi)^/release-note-action-required\s*$`)
 )
 
 func init() {
-	plugins.RegisterIssueCommentHandler(pluginName, handleIssueComment)
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest)
```
Comment (spxtr) [inline prow/plugins/releasenote/releasenote.go]: Is it possible to use issue events instead of PR events? IIRC issue events have the labels in the webhook, so we would save some API tokens.

**Comment by kargakis** [comment]
> Quite a few people use this functionality, and I like that I can say /release-note-none in a comment if I forget to say NONE in the PR body. I suggest leaving at least the /release-note-none command. We will also need to broadcast that we are changing these commands.

Yeah, I have struggled with the NONE field before and the command feels like a no-brainer to use.

**Comment by cjwagner** [comment]
@spxtr @kargakis If I include the fix for #4341 and keep the `/release-note` commands, the label specified by the commands can differ from the label specified by the `release-note` block in the PR body. This would cause weird behavior when any part of the PR body is updated because the `release-note` block will be checked again and the appropriate label will be applied even if a `/release-note` command has been previously made.
To maintain a single source of truth the `/release-note` commands would need to be updated to change the `release-note` block of the PR body text in addition to modifying the PR labels (e.g. `/release-note-none` would need to change the `release-note` block to the string "none") **OR** the labels could be used as the source of truth in which case #4341 cannot be fixed.

File: prow/github/fakegithub/fakegithub.go [hunk 3]
```diff
@@ -105,6 +105,18 @@ func (f *FakeClient) DeleteComment(owner, repo string, ID int) error {
 	return fmt.Errorf("could not find issue comment %d", ID)
 }
 
+func (f *FakeClient) DeleteStaleComments(org, repo string, number int, isStale func(github.IssueComment) bool) error {
+	comments, _ := f.ListIssueComments(org, repo, number)
```
Comment (stevekuznetsov) [inline prow/github/fakegithub/fakegithub.go]: The mungers have this abstracted so we only list once, right? Do we intend for many plugins to have this functionality? Should we attempt to build a similar abstraction here?

File: prow/plugins/releasenote/releasenote.go [hunk 4]
```diff
@@ -29,98 +31,272 @@ import (
...
+	// Clean up old comments.
+	// If the PR must follow the process and hasn't yet completed the process, don't remove comments.
+	if prMustFollowRelNoteProcess(gc, log, pr, prLabels, false) && !releaseNoteAlreadyAdded(prLabels) {
+		return nil
+	}
+	botName, err := gc.BotName()
 	if err != nil {
 		return err
 	}
+	return gc.DeleteStaleComments(
```
[the remark this answers is not in the document]
Comment (cjwagner) [reply prow/plugins/releasenote/releasenote.go]: The `handlePR` function returns before reaching this point unless the PR doesn't need to follow the release note process or has already completed the process.

**Comment by rmmh** [comment]
I think there are a few distinct states to consider:

1. release note body "NONE" => label `release-note-none`
2. release note body (non-NONE) => label `release-note`
3. empty (or missing) release note body and /release-note-none command => label `release-note-none`
4. empty (or missing) release note body => label `do-not-merge/release-note-label-needed`

Apply the first rule that matches on an PR edit event or when a comment has `/release-note-none`. Remove the other release-note commands, or make them respond with usage-- (`/release-note whatever` replies with `edit the release note in the PR body to set a release note`).

I think this is intuitive and less complex than having the bot edit the PR body.

**Comment by spxtr** [comment]
https://github.com/kubernetes/test-infra/pull/4718#issuecomment-332347895 should work, although it means we need to list issue comments every time we get to step 3. That should be fine, however.

File: prow/github/fakegithub/fakegithub.go (hunk 3 above)
Comment (cjwagner) [reply prow/github/fakegithub/fakegithub.go]: Thats correct, and something similar will probably be necessary for Prow, but lets wait until there is another plugin that requires stale comment deletion before implementing that.

File: prow/plugins/releasenote/releasenote.go (hunk 1 above)
Comment (cjwagner) [reply prow/plugins/releasenote/releasenote.go]: Done.

File: prow/plugins/releasenote/releasenote.go (hunk 2 above)
Comment (cjwagner) [reply prow/plugins/releasenote/releasenote.go]: Issue events would include the labels, but they wouldn't include the PR itself which is needed to check the branch that the PR is for. So switching to issue event instead of PR events would use the same number of API tokens.

**Comment by cjwagner** [comment]
Comments addressed. PTAL

File: prow/plugins/releasenote/releasenote.go [hunk 5]
```diff
@@ -114,6 +115,13 @@ func handleComment(gc githubClient, log *logrus.Entry, ic github.IssueCommentEve
 		return nil
 	}
 
+	// Emit deprecation warning for /release-note and /release-note-action-required.
+	if nl == releaseNote || nl == releaseNoteActionRequired {
+		format := "the `/%s` and `/%s` commands have been deprecated.\nPlease edit the `release-note` block in the PR body text to include the release note. If the release note requires additional action include the string `action required` in the release note. For example:\n````\n```release-note\nSome release note with action required.\n```\n````"
```
Comment (rmmh) [inline prow/plugins/releasenote/releasenote.go]: Oh, so that's how you quote triple quotes!
````
```release-note
Some release note with action required.
```
````

Comment (cjwagner) [reply prow/plugins/releasenote/releasenote.go]: Yep. You can do it by bracketing with 2 backticks instead of 4 as well, but that version doesn't display multiple lines.

File: prow/plugins/releasenote/releasenote.go [hunk 6]
```diff
@@ -123,25 +131,29 @@ func handleComment(gc githubClient, log *logrus.Entry, ic github.IssueCommentEve
...
 		return gc.CreateComment(org, repo, number, plugins.FormatICResponse(ic.Comment, resp))
 	}
 
-	// Add the requested label if necessary.
-	var errs []error
-	if !ic.Issue.HasLabel(nl) {
-		log.Infof("Adding %s label.", nl)
-		if err := gc.AddLabel(org, repo, number, nl); err != nil {
-			errs = append(errs, err)
+	// Add the releaseNoteNone label if the release-note block is none or empty.
```
Comment (rmmh) [inline prow/plugins/releasenote/releasenote.go]: Is this comment correct? It looks like it's adding a comment if there's release note or releasenoteactionrequired?

Comment (cjwagner) [reply prow/plugins/releasenote/releasenote.go]: Its correct. I changed it to 
"Don't allow the /release-note-none command if the release-note block contains a valid release note."
to be more clear.

**Comment by xiangpengzhao** [comment]
Update PR template and docs:
https://github.com/kubernetes/kubernetes/pull/54114
https://github.com/kubernetes/community/pull/1221

@cjwagner PTAL
